File: Linear Control Course/Chapter1/Lesson2/Chapter1_Lesson2.java
public class FirstOrderFeedbackExample {
    public static void main(String[] args) {
        double a = 1.0;
        double b = 1.0;
        double K = 2.0;
        double h = 0.01;
        int    N = 1000;
        double r = 1.0;

        double[] yOpen = new double[N + 1];
        double[] yClosed = new double[N + 1];

        double uOpen = a / b * r;

        for (int k = 0; k < N; ++k) {
            // Open-loop update
            yOpen[k + 1] = yOpen[k] +
                    h * (-a * yOpen[k] + b * uOpen);

            // Closed-loop update
            double eK = r - yClosed[k];
            double uClosed = K * eK;
            yClosed[k + 1] = yClosed[k] +
                    h * (-a * yClosed[k] + b * uClosed);
        }

        System.out.println("k,time,yOpen,yClosed");
        for (int k = 0; k <= N; ++k) {
            double t = k * h;
            System.out.println(
                k + "," + t + "," + yOpen[k] + "," + yClosed[k]
            );
        }
    }
}

File: Linear_Control_Course/Chapter1/Lesson2/Chapter1_Lesson2.py
import numpy as np
import matplotlib.pyplot as plt

# Plant parameters
a = 1.0
b = 1.0

# Controller gain
K = 2.0

# Discretization
h = 0.01      # time step [s]
N = 1000      # number of steps
r = 1.0       # constant reference

t = np.linspace(0.0, N * h, N + 1)

y_open = np.zeros(N + 1)
y_closed = np.zeros(N + 1)

# Choose open-loop input to roughly achieve y ≈ 1 at steady state
u_open = a / b * r

for k in range(N):
    # Open-loop update
    y_open[k + 1] = y_open[k] + h * (-a * y_open[k] + b * u_open)

    # Closed-loop update
    e_k = r - y_closed[k]
    u_closed = K * e_k
    y_closed[k + 1] = y_closed[k] + h * (-a * y_closed[k] + b * u_closed)

# Plot for visualization
plt.plot(t, y_open, label="open-loop")
plt.plot(t, y_closed, label="closed-loop")
plt.axhline(r, linestyle="--", label="reference")
plt.xlabel("time [s]")
plt.ylabel("output y(t)")
plt.grid(True)
plt.legend()
plt.show()
